### PR TITLE
Implement captcha modal in AuthUI v2

### PR DESCRIPTION
**Pull request by pkong-ds**
## What's in this PR?
Implement captcha guard in
- [x] signup flow
  - [x] identify
    - [x] username
    - [x] email
    - [x] phone
- [x] promote flow
  - [x] identify
    - [x] username
    - [x] email
    - [x] phone
- [x] login flow
  - [x] identify
    - [x] loginid
  - [x] authenticate
    - [x] password
    - [ ] oobotp (in separate page)
- [x] signup_login flow
  - [x] switch flow success
  - [x] identify
    - [x] loginid
- [x] account_recovery_flow
  - [x] identify
  
## Preview 

https://github.com/user-attachments/assets/b4568a15-abf4-47b5-9dd2-6348f641c0a6

## Notes on 502 error

If you noticed a 502 gate error in preview, it's because of header too large (>4kb) again.

1. grecaptcha has an embedded `<textarea />` input, which store the very long grecaptcha token
2. I also have an input field `x_bot_protection_provider_token`, store the very long grecaptcha token _again_
3. Empty phone input will return an error, **with entire form input returned**
4. Each token cost ~ 500b - 1kb, so the response headers with `web_err` in set cookie exceeded 4kb easily.

<details>
<summary> Sample repsonse Header (4372 bytes) </summary>

```
Cache-Control: no-store
Content-Security-Policy: default-src 'self'; script-src 'strict-dynamic' 'nonce-SASWCQCQ9NNC3NZRB68QDMXM7CTZHKDQ' www.googletagmanager.com eu-assets.i.posthog.com challenges.cloudflare.com www.google.com https://browser.sentry-cdn.com 'self'; frame-src www.googletagmanager.com challenges.cloudflare.com www.google.com 'self'; font-src cdnjs.cloudflare.com static2.sharepointonline.com fonts.googleapis.com fonts.gstatic.com 'self'; style-src 'unsafe-inline' cdnjs.cloudflare.com www.googletagmanager.com fonts.googleapis.com 'self'; img-src http: https: data: 'self'; object-src 'none'; base-uri 'none'; connect-src 'self' https://www.google-analytics.com ws://accounts.portal.localhost:3000 wss://accounts.portal.localhost:3000; block-all-mixed-content; frame-ancestors 'none'
Location: /login?q_login_id_input_type=phone
Permissions-Policy: accelerometer=(), ambient-light-sensor=(), autoplay=*, battery=(), bluetooth=(), browsing-topics=(), camera=(), display-capture=(), document-domain=(), encrypted-media=(), execution-while-not-rendered=*, execution-while-out-of-viewport=*, fullscreen=*, gamepad=(), geolocation=(), gyroscope=(), hid=(), identity-credentials-get=(), idle-detection=(), local-fonts=(), magnetometer=(), microphone=(), midi=(), otp-credentials=(), payment=(), picture-in-picture=(), publickey-credentials-create=(self), publickey-credentials-get=(self), screen-wake-lock=(), serial=(), speaker-selection=(), storage-access=(), usb=(), web-share=(), window-management=(), xr-spatial-tracking=()
Pragma: no-cache
Set-Cookie: debug_csrf_same_site_omit=exists; Path=/; Domain=portal.localhost; Max-Age=1200; HttpOnly
Set-Cookie: debug_csrf_same_site_none=exists; Path=/; Domain=portal.localhost; Max-Age=1200; HttpOnly
Set-Cookie: debug_csrf_same_site_lax=exists; Path=/; Domain=portal.localhost; Max-Age=1200; HttpOnly; SameSite=Lax
Set-Cookie: debug_csrf_same_site_strict=exists; Path=/; Domain=portal.localhost; Max-Age=1200; HttpOnly; SameSite=Strict
Set-Cookie: web_err=eyJGb3JtIjp7ImctcmVjYXB0Y2hhLXJlc3BvbnNlIjpbIjAzQUZjV2VBN05sc2s5WXNqb0tLYXUwLUEzNE5Pa2tUUXdQRDJKc29oRmRacG92TDAxRS1EYUtwNk5wNTlOUDNfTjRZVXhyaF9PSkwyc3hHekgwakNVQks3LXV4eWh0TDhIZXRFdUpKWGtKbHAwcWN4UG0zVzFoY2s4Qnh2QTJYY2pOVVZpaS1TYndKVEdMRkVaTjBvejJNTEZZRDczem1PT2pfWFBTTVdtMkNVZE5PajgxSlpEQnVxOFlaVGNRTFZsSFlMdnRFbGR3OEo0b2dNazUxWUIxTkRhdTVvQWllWnBNaWJtXzNwOXFZQ2JfbENSblJEZ0NXTXAzUzhfZ0RQM3Y5SlBNUHZ2Yll0WWtSQ001SGFlV1V2aUR0dFVWWE9HdUh6OTBTNndwUWNhUEdmVnZ3OFNvZ19RdVNXX193ODRya1BmSkp4NGluS05DTV9MRjZ2bHN2a2NoanRJVmxaNXJxSXhRZV9kek5tSFpjczhaMHh2SDBpcHZnc3FNOW55V0V1MUdXMi1zcUhUUXV1Si1ta0xiV0RaU3JSUGREa05TbjhEUGFDZ3lrVllBNUgzY3VEZEJ4S2NtemFmUkt4bmZTdW5Ebjc4S1dyTml0TEU3N0haejIwSTZVUkxvOWVzdld0bmpZUUpHNVBVZlphQ3ZVWmVBc3pXb0dWM250T3FyVk1PM1dtTVR5eXFwQTY3MmxWUWh5bVBIamNiVXZLY2Z3Il0sImdvcmlsbGEuY3NyZi5Ub2tlbiI6WyIwS0QzSDlQdEZaMC9VTVpaazQweko0RE1wYmluV0Zrd3RhNTNzWHlQeDZuTTFCck54RWthMjVTam4ra01ZVGFxZWY3UVhvdkxSU0YvQ3hXUTFPU1pTUT09Il0sInFfbG9naW5faWRfaW5wdXRfdHlwZSI6WyJwaG9uZSJdLCJ4X2FjdGlvbiI6WyJsb2dpbl9pZCJdLCJ4X2JvdF9wcm90ZWN0aW9uX3Byb3ZpZGVyX3Jlc3BvbnNlIjpbIjAzQUZjV2VBN05sc2s5WXNqb0tLYXUwLUEzNE5Pa2tUUXdQRDJKc29oRmRacG92TDAxRS1EYUtwNk5wNTlOUDNfTjRZVXhyaF9PSkwyc3hHekgwakNVQks3LXV4eWh0TDhIZXRFdUpKWGtKbHAwcWN4UG0zVzFoY2s4Qnh2QTJYY2pOVVZpaS1TYndKVEdMRkVaTjBvejJNTEZZRDczem1PT2pfWFBTTVdtMkNVZE5PajgxSlpEQnVxOFlaVGNRTFZsSFlMdnRFbGR3OEo0b2dNazUxWUIxTkRhdTVvQWllWnBNaWJtXzNwOXFZQ2JfbENSblJEZ0NXTXAzUzhfZ0RQM3Y5SlBNUHZ2Yll0WWtSQ001SGFlV1V2aUR0dFVWWE9HdUh6OTBTNndwUWNhUEdmVnZ3OFNvZ19RdVNXX193ODRya1BmSkp4NGluS05DTV9MRjZ2bHN2a2NoanRJVmxaNXJxSXhRZV9kek5tSFpjczhaMHh2SDBpcHZnc3FNOW55V0V1MUdXMi1zcUhUUXV1Si1ta0xiV0RaU3JSUGREa05TbjhEUGFDZ3lrVllBNUgzY3VEZEJ4S2NtemFmUkt4bmZTdW5Ebjc4S1dyTml0TEU3N0haejIwSTZVUkxvOWVzdld0bmpZUUpHNVBVZlphQ3ZVWmVBc3pXb0dWM250T3FyVk1PM1dtTVR5eXFwQTY3MmxWUWh5bVBIamNiVXZLY2Z3Il0sInhfYm90X3Byb3RlY3Rpb25fcHJvdmlkZXJfdHlwZSI6WyJyZWNhcHRjaGF2MiJdLCJ4X2xvZ2luX2lkIjpbIis4NTIiXSwieF9sb2dpbl9pZF9pbnB1dF90eXBlIjpbInBob25lIl19LCJFcnJvciI6eyJuYW1lIjoiTm90Rm91bmQiLCJyZWFzb24iOiJVc2VyTm90Rm91bmQiLCJtZXNzYWdlIjoidXNlciBub3QgZm91bmQiLCJjb2RlIjo0MDQsImluZm8iOnsiRmxvd1R5cGUiOiJsb2dpbiIsIklkZW50aXR5VHlwZUluY29taW5nIjoibG9naW5faWQiLCJMb2dpbklEVHlwZUluY29taW5nIjoiIn19fQ; Path=/; Domain=portal.localhost; HttpOnly; SameSite=Lax
Vary: Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
Date: Thu, 18 Jul 2024 18:14:11 GMT
Content-Length: 0
```

</details>

<details>

<summary>  `base64 -d web_err`   </summary>

```json5
{
  "Form": {
    "g-recaptcha-response": [
      "03AFcWeA7Nlsk9YsjoKKau0-A34NOkkTQwPD2JsohFdZpovL01E-DaKp6Np59NP3_N4YUxrh_OJL2sxGzH0jCUBK7-uxyhtL8HetEuJJXkJlp0qcxPm3W1hck8BxvA2XcjNUVii-SbwJTGLFEZN0oz2MLFYD73zmOOj_XPSMWm2CUdNOj81JZDBuq8YZTcQLVlHYLvtEldw8J4ogMk51YB1NDau5oAieZpMibm_3p9qYCb_lCRnRDgCWMp3S8_gDP3v9JPMPvvbYtYkRCM5HaeWUviDttUVXOGuHz90S6wpQcaPGfVvw8Sog_QuSW__w84rkPfJJx4inKNCM_LF6vlsvkchjtIVlZ5rqIxQe_dzNmHZcs8Z0xvH0ipvgsqM9nyWEu1GW2-sqHTQuuJ-mkLbWDZSrRPdDkNSn8DPaCgykVYA5H3cuDdBxKcmzafRKxnfSunDn78KWrNitLE77HZz20I6URLo9esvWtnjYQJG5PUfZaCvUZeAszWoGV3ntOqrVMO3WmMTyyqpA672lVQhymPHjcbUvKcfw"
    ],
    "gorilla.csrf.Token": [
      "0KD3H9PtFZ0/UMZZk40zJ4DMpbinWFkwta53sXyPx6nM1BrNxEka25Sjn+kMYTaqef7QXovLRSF/CxWQ1OSZSQ=="
    ],
    "q_login_id_input_type": ["phone"],
    "x_action": ["login_id"],
    "x_bot_protection_provider_response": [
      "03AFcWeA7Nlsk9YsjoKKau0-A34NOkkTQwPD2JsohFdZpovL01E-DaKp6Np59NP3_N4YUxrh_OJL2sxGzH0jCUBK7-uxyhtL8HetEuJJXkJlp0qcxPm3W1hck8BxvA2XcjNUVii-SbwJTGLFEZN0oz2MLFYD73zmOOj_XPSMWm2CUdNOj81JZDBuq8YZTcQLVlHYLvtEldw8J4ogMk51YB1NDau5oAieZpMibm_3p9qYCb_lCRnRDgCWMp3S8_gDP3v9JPMPvvbYtYkRCM5HaeWUviDttUVXOGuHz90S6wpQcaPGfVvw8Sog_QuSW__w84rkPfJJx4inKNCM_LF6vlsvkchjtIVlZ5rqIxQe_dzNmHZcs8Z0xvH0ipvgsqM9nyWEu1GW2-sqHTQuuJ-mkLbWDZSrRPdDkNSn8DPaCgykVYA5H3cuDdBxKcmzafRKxnfSunDn78KWrNitLE77HZz20I6URLo9esvWtnjYQJG5PUfZaCvUZeAszWoGV3ntOqrVMO3WmMTyyqpA672lVQhymPHjcbUvKcfw"
    ],
    "x_bot_protection_provider_type": ["recaptchav2"],
    "x_login_id": ["+852"],
    "x_login_id_input_type": ["phone"]
  },
  "Error": {
    "name": "NotFound",
    "reason": "UserNotFound",
    "message": "user not found",
    "code": 404,
    "info": {
      "FlowType": "login",
      "IdentityTypeIncoming": "login_id",
      "LoginIDTypeIncoming": ""
    }
  }
}

```

</details>

Related [Linear Issue](https://linear.app/authgear/issue/DEV-1537/review-response-headers-content-security-policy)

Will look into more tmr how to workaround this

Note - [cloudflare supports enabling/disabling this](https://developers.cloudflare.com/turnstile/get-started/client-side-rendering/#configurations), but not recaptcha :sadfrog: 

> `response_field` - A boolean that controls if an input element with the response token is created, defaults to true.

### CODE REVIEW

**Comment by pkong-ds** [comment]
To make this PR smaller, maybe we can do one PR for each flow type/step. i.e.

PR 1 - login identify
PR 2 - login authentication
PR 3 - signup identify 
...

See what daic prefer 🙏

**Comment by pkong-ds** [comment]
Added fade-in fade-out

https://github.com/user-attachments/assets/029dc7b9-5683-42df-a9c3-62bbab6da4ae

**Comment by pkong-ds** [comment]
This PR feature logic is complete. pending modal design and animation. Suggest review first 🙏

File: resources/authgear/templates/en/web/authflowv2/login.html [hunk 1]
```diff
@@ -91,6 +102,9 @@ <h2 class="screen-description">
       class="flex flex-col gap-4 mt-8"
       method="post"
       novalidate
+      {{ if (or (and $show_phone_input $show_captcha_phone) (and $show_text_input (or $show_captcha_email $show_captcha_username)))}}
```
Comment (louischan-oursky) [inline resources/authgear/templates/en/web/authflowv2/login.html:105]: The condition can be simplified into `if (or $show_captcha_phone $show_captcha_email $show_captcha_username)`.

Rationale: You have checked `$show_phone_input` or `$show_text_input` when you set `$show_captcha_*` to true. So `(and $show_phone_input $show_captcha_*` is redundant.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by pkong-ds** [comment]
@louischan-oursky updated, thanks!

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**